docs/design-documents/hal/0003-explicit-pinmap-extension.md

```diff
@@ -0,0 +1,302 @@
+1. [System architecture and high-level design](#system-architecture-and-high-level-design).
+    1. [Explicit pinmap mechanism](#explicit-pinmap-mechanism).
+    1. [How much we can save](#how-much-we-can-save).
+1. [Detailed design](#detailed-design).
+    1. [Pinmap types](#pinmap-types).
+    1. [HAL API changes](#hal-api-changes).
+    1. [Drivers API changes](#drivers-api-changes).
+    1. [`constexpr` utility functions](#constexpr-utility-functions).
+    1. [`Example usage](#constexpr-utility-functions).
+
+# Introduction
+
+Making use of the pinmap tables optional by allowing the peripheral configuration (pin/periheral/function) to be explicitly specified in the HAL API function.
+
+### Overview and background
+
+HAL APIs making use of pins take these pins in their constructor and use those pins to lookup which peripheral/function to use. The process of looking up the peripheral/function requires there to be a pinmap table which maps pins to peripherals/functions. This pinmap table takes up ROM which could be saved if the pinmap wasn't used.
+The goal is to provide additional HAL API/constructors which takes pinmap as a parameter where pin/peripheral/function is specified explicitly and there is no need to use the pinmap tables.
+This extension should give the following savings:
+- removed pinmap tables,
+- removed `hal\mbed_pinmap_common library` (required for pin lookup mechanism),
+- reduced HAL driver code.
+
+### Requirements and assumptions
+
+1. Provide types which will hold explicit pinmaps for peripherals(`PWM`, `AnalogIn`, `AnalogOut`, `SPI`, `I2C`, `UART`, `QSPI`, `CAN`).
+2. Provide `xxx_init_direct(xxx_t *obj, explicit_pinmap_t *)` functions to HAL API (these functions will not use pinmap tables).
+3. Provide additional constructors in drivers layer which will use the `xxx_init_direct(xxx_t *obj, explicit_pinmap_t*)` HAL functions.
+4. Provide default weak implementations of `xxx_init_direct(explicit_pinmap_t *)` functions. These functions will call standard `xxx_init(xxx_t *obj, PinName, ...)` function (backward compatibility for targets which do not support explicit pinmap mechanism).
+5. Provide `constexpr` utility functions to lookup for pin mapping in compile time (requires C++14).
+6. Initialize console using explicit pinmap mechanism, so `hal\mbed_pinmap_common library` is not needed and can be removed.
+7. Modify FPGA tests to verify also `xxx_init_direct(xxx_t *obj, explicit_pinmap_t*)` APIs.
+
+# System architecture and high-level design
+
+### Explicit pinmap mechanism
+
+The explicit pinmap mechanism with backward compatibility is shown below on PWM peripheral example.
+
+![alt text](img/explicit_pinmap.png "Explicit pinmap model")
+
+For targets which do not provide explicit pinmap support standard initialization will be performed (which uses pinmap tables) even if direct API is selected.
+
+### How much we can save
+
+Tables below show how much ROM we can save using explicit pinmap for each peripheral:
```

Comment:
> Also the difference between the basic version of the extension and the 'constexpr' is 0 or close to 0, would be good to explain why we do it. Could be it's somewhere below and I just didn't get to it.

That was one of my last thoughts on the PR - there is no fundamental point to having both the basic extension, as both it and the `constexpr` form should generate exactly the same code.

I believe in the final form of the PR, we've actually ended up so that targets only implement `init_direct` themselves if *also* exposing the pins. That was to ensure the new `constexpr` form is fully portable, and is always usable

* Old target: default weak get_pinmap fills in dummy pinmap structure, default weak `init_direct` implementation calls standard HAL `init`.
* New target (`EXPLICIT_PINMAP_READY`): implements `get_pinmap` and `init_direct`.

So there is no remaining justification for the basic form.  Except ARMC5, I guess.

Reply:
> Also the difference between the basic version of the extension and the 'constexpr' is 0 or close to 0, would be good to explain why we do it. Could be it's somewhere below and I just didn't get to it.

Ok, I will add the explanation. We expect that using `constexpr` utility functions should not give regression in the memory savings (otherwise adding this feature is pointless). I added this row to show that we don't have regression.

> Note on methodology would be also good, what are we measuring? Simple example with only one peripheral used? Cool, lets have a short description. Also would be good to have complex example using I2C, SPI, UART, GPIO and AnalogIn.

I'm still adding support for the remaining golden targets. I'm planning to add remaining targets in the separate PR since PR https://github.com/ARMmbed/mbed-os/pull/11399 is big enough. When this is ready I will change the tables and add mem stats for all supported targtes/peripherals together with the examples.